app/src/main/java/com/github/jorgecastillo/kotlinandroid/functional/Future.kt

```diff
@@ -1,35 +1,68 @@
 package com.github.jorgecastillo.kotlinandroid.functional
 
+import kategory.*
 import kotlinx.coroutines.experimental.CommonPool
 import kotlinx.coroutines.experimental.Deferred
 import kotlinx.coroutines.experimental.android.UI
 import kotlinx.coroutines.experimental.async
 import kotlinx.coroutines.experimental.launch
 
+typealias FutureKind<A> = HK<Future.F, A>
+
+fun <A> FutureKind<A>.ev(): Future<A> =
+        this as Future<A>
+
 /**
  * Basic future implementation to achieve asynchronicity based on Kotlin coroutines.
  */
-class Future<T> {
+class Future<T> : FutureKind<T> {
+
+    class F private constructor()
+
+    private val deferred: Deferred<T>
+
+    private constructor(deferred: Deferred<T>) {
+        this.deferred = deferred
+    }
+
+    constructor(f: () -> T) : this(async(CommonPool) { f() })
 
-  private val deferred: Deferred<T>
+    fun <X> map(f: (T) -> X): Future<X> {
+        return Future(async(CommonPool) { f(deferred.await()) })
+    }
+
+    fun <X> flatMap(f: (T) -> Future<X>): Future<X> {
+        return Future(async(CommonPool) { f(deferred.await()).deferred.await() })
+    }
+
+    fun <B> zip(fb: Future<B>): Future<Tuple2<T, B>> =
+            flatMap { a -> fb.map { b -> Tuple2(a, b) } }
+
+    fun onComplete(f: (T) -> Unit) {
+        launch(UI) {
+            f(deferred.await())
+        }
+    }
 
-  private constructor(deferred: Deferred<T>) {
-    this.deferred = deferred
-  }
+    companion object : Monad<F> {
+        override fun <A, B> flatMap(fa: FutureKind<A>, f: (A) -> FutureKind<B>): Future<B> =
```

Comment:
Answered above, these types should not live in this project as they are general purpose https://github.com/JorgeCastilloPrz/KotlinAndroidFunctional/pull/2#discussion_r127655693

As for FutureKind, it needs to be that to remain compatible with the Monad interface since Kotlin does not support HKT. In whatever case you can pass a Future to it. It's also common practice in FP in Scala and langs with companions to have the companion be the instance that implements the typeclasses so one can just use the companion without invoking typeclasses.